spring.go

```diff
@@ -0,0 +1,405 @@
+			break
+		}
+		child := value.FieldByName(field.Name)
+		inject(input, child, bindingMethod, occasion, level+1)
+	}
+}
+
+func createAndSetDefaultValue(value *reflect.Value, fieldIndex int) (target reflect.Value) {
+	fv := value.Field(fieldIndex)
+	if ignored(fv.Type()) {
+		return
+	}
+	fvk := fv.Kind()
+	if fvk == reflect.Struct {
+		target = fv
+		return
+	}
+	if fvk == reflect.Interface {
+		id := fv.Type().Name() + "global"
+		if ct, exist := container[id]; exist {
+			cv := reflect.New(ct)
+			if fv.CanSet() && cv.IsValid() {
+				fv.Set(cv)
+				target = cv
+				return
+			}
+		}
+	}
+	if !fv.CanSet() {
+		return
+	}
+	ft := fv.Type()
+	if ft.Kind() != reflect.Ptr || !fv.IsNil() {
+		return
+	}
+
+	for ft.Kind() == reflect.Ptr {
+		ft = ft.Elem()
+		fv = fv.Elem()
+	}
+	if !fv.CanSet() || ft.Kind() != reflect.Struct {
+		return
+	}
+	target = reflect.New(ft).Elem()
+	fv.Set(target)
+	return
+}
+
+func ignored(t reflect.Type) bool {
+	for ; t.Kind() == reflect.Ptr; t = t.Elem() {
+	}
+	for _, tt := range ignoredTypes {
+		if t == tt {
+			return true
+		}
+	}
+	return false
+}
+
+func getUnderlayType(value reflect.Value) (underlayType reflect.Type, underlayValue reflect.Value, injectable bool) {
+	if !value.IsValid() {
+		return
+	}
+	t := value.Type()
+	k := t.Kind()
+	for {
+		if k == reflect.Struct {
+			underlayType = t
+			underlayValue = value
+			injectable = true
+			return
+		} else if k == reflect.Ptr {
+			t = t.Elem()
+			value = value.Elem()
+			k = t.Kind()
+		} else {
+			break
+		}
+	}
+	return
+}
+
+func getBeanConfig(tagValue, bindingMethod, occasion string) (beanName string, errorHandler ErrorResolver, notForThisMethod bool) {
+	configs := strings.SplitN(tagValue, ";", -1)
+	occ := PreHandling
+	for _, config := range configs {
+		c := strings.TrimSpace(config)
+		if o := occasionPattern.FindStringSubmatch(c); o != nil {
+			occ = strings.TrimSpace(o[1])
+			if occ != occasion {
+				notForThisMethod = true
+				return
+			}
+		}
+		if bm := bindMethodPattern.FindStringSubmatch(c); bm != nil {
+			rawBM := bm[1]
+			bindingMethods := strings.SplitN(rawBM, ",", -1)
+			found := false
+			for _, bmm := range bindingMethods {
+				if strings.TrimSpace(bmm) == bindingMethod {
+					found = true
+					break
+				}
+			}
+			if !found {
+				notForThisMethod = true
+				return
+			}
+		}
+		if bn := beanNamePattern.FindStringSubmatch(c); bn != nil {
+			beanName = bn[1]
+			continue
+		}
+		if eh := errorHandlerPattern.FindStringSubmatch(c); eh != nil {
+			ehName := eh[1]
+			if ehh, exist := errorHandlers[ehName]; exist {
+				errorHandler = ehh
+				continue
+			}
+			panic(ErrorHandlerNotFound)
+		}
+	}
+	if occ != occasion {
+		notForThisMethod = true
+		return
+	}
+	if errorHandler == nil {
+		errorHandler = nilOnError
+	}
+	return
+}
+
+func injectOneField(field *reflect.StructField, value *reflect.Value, fieldIndex int, input *context.Context, bindingMethod, occasion string, level int) (breakOperation bool) {
```

Comment:
warning: cyclomatic complexity 16 of function injectOneField() is high (> 15) (gocyclo)